src/engine/strat_engine/keys.rs

```diff
@@ -0,0 +1,428 @@
+pub fn read_key(
+    key_desc: &KeyDescription,
+) -> StratisResult<(Option<(KeySerial, SizedKeyMemory)>, KeySerial)> {
+    let keyring_id = get_persistent_keyring()?;
+
+    let key_id_option = search_key(keyring_id, key_desc)?;
+    let key_id = if let Some(ki) = key_id_option {
+        ki
+    } else {
+        return Ok((None, keyring_id));
+    };
+
+    let mut key_buffer = SafeMemHandle::alloc(MAX_STRATIS_PASS_SIZE)?;
+    let mut_ref = key_buffer.as_mut();
+
+    // Read key from keyring
+    match unsafe {
+        syscall(
+            SYS_keyctl,
+            libc::KEYCTL_READ,
+            key_id,
+            mut_ref.as_mut_ptr(),
+            mut_ref.len(),
+        )
+    } {
+        i if i < 0 => Err(io::Error::last_os_error().into()),
+        i => Ok((
+            Some((
+                key_id as KeySerial,
+                SizedKeyMemory::new(key_buffer, i as usize),
+            )),
+            keyring_id,
+        )),
+    }
+}
+
+/// Reset the key data attached to the provided key description if the new key data
+/// is different from the old key data.
+// Precondition: The key description is already present in the keyring.
+fn reset_key(
+    key_id: KeySerial,
+    old_key_data: SizedKeyMemory,
+    new_key_data: SizedKeyMemory,
+) -> StratisResult<bool> {
+    if old_key_data.as_ref() == new_key_data.as_ref() {
+        Ok(false)
+    } else {
+        // Update the existing key data
+        let update_result = unsafe {
+            syscall(
+                SYS_keyctl,
+                libc::KEYCTL_UPDATE,
+                key_id,
+                new_key_data.as_ref().as_ptr(),
+                new_key_data.as_ref().len(),
+            )
+        };
+        if update_result < 0 {
+            Err(io::Error::last_os_error().into())
+        } else {
+            Ok(true)
+        }
+    }
+}
+
+/// Add the key to the given keyring under with the provided key description.
+// Precondition: The key description was not already present.
+fn set_key(
+    key_desc: &KeyDescription,
+    key_data: SizedKeyMemory,
+    keyring_id: KeySerial,
+) -> StratisResult<()> {
+    let key_desc_cstring = CString::new(key_desc.to_string()).map_err(|_| {
+        StratisError::Engine(
+            ErrorEnum::Invalid,
+            "Invalid key description provided".to_string(),
+        )
+    })?;
+    // Add a key to the kernel keyring
+    if unsafe {
+        libc::syscall(
+            SYS_add_key,
+            concat!("user", "\0").as_ptr(),
+            key_desc_cstring.as_ptr(),
+            key_data.as_ref().as_ptr(),
+            key_data.as_ref().len(),
+            keyring_id,
+        )
+    } < 0
+    {
+        Err(io::Error::last_os_error().into())
+    } else {
+        Ok(())
+    }
+}
+
+/// Perform an idempotent add of the given key data with the given key description.
+///
+/// Successful return values:
+/// * `Ok(CreateAction::Identity)`: The key was already in the keyring with the
```

Comment:
Why can't you use ```RenameAction``` instead of ```CreateAction``` here? You could rename ```RenameAction``` to something more general, but it seems like it captures the sematics better. I recall that we discussed a possibility of improving ```RenameAction``` so it may seem a bit unsatisfactory to use, but it exists and it does capture all the cases.

Reply:
Structurally, I think `Rename` provides what we want here, but these actions are very specific to what action we're performing. There's no condition where there's no source for the create action because create actions don't have a source and target, only a target. If we rename `NoSource` to something else, we lose the meaning of what it's supposed to indicate which is very specific to renaming. I worry that we'll bump eventually introduce another bug related to  overloading types if we try to make it too general.

If you don't think the current solution is workable, I see the two most workable solutions here being creating a new type for `Create` actions that we only use here (something like `CreateUpdateAction`, and we may end up using it in the future as this kind of pattern is common in idempotent frameworks like Ansible) or extending the `CreateAction` type and reworking the other places that it's used. In either solution, we could potentially add a `Updated`/`Changed` variant to indicate that the identifier remained the same but the linked data was changed.

Reply:
> Structurally, I think `Rename` provides what we want here, but these actions are very specific to what action we're performing. There's no condition where there's no source for the create action because create actions don't have a source and target, only a target. If we rename `NoSource` to something else, we lose the meaning of what it's supposed to indicate which is very specific to renaming. I worry that we'll eventually introduce another bug related to overloading types if we try to make it too general.
> 
> If you don't think the current solution is workable, I see the two most workable solutions here being creating a new type for `Create` actions that we only use here (something like `CreateUpdateAction`, and we may end up using it in the future as this kind of pattern is common in idempotent frameworks like Ansible) or extending the `CreateAction` type and reworking the other places that it's used. In either solution, we could potentially add a `Updated`/`Changed` variant to indicate that the identifier remained the same but the linked data was changed.

You have a ```SetCreateAction```, why not make a ```MappingCreateAction``` with constructors:
```Identity```, ```ValueChanged```, ```Created``` or something like that?